src/libmongoc/src/mongoc/mongoc-util.c

```diff
@@ -626,3 +627,154 @@ _mongoc_getenv (const char *name)
 
 #endif
 }
+
+/* Nearly Divisionless (Algorithm 5): https://arxiv.org/abs/1805.10941 */
+static uint32_t
+_mongoc_rand_nduid32 (uint32_t s, uint32_t (*rand32) (void))
+{
+   const uint64_t limit = UINT32_MAX; /* 2^L */
+   uint64_t x, m, l;
+
+   x = rand32 ();
+   m = x * s;
+   l = m % limit;
+
+   if (l < s) {
+      const uint64_t t = (limit - s) % s;
+
+      while (l < t) {
+         x = rand32 ();
+         m = x * s;
+         l = m % limit;
+      }
+   }
+
+   return (uint32_t) (m / limit);
+}
+
+/* Java Algorithm (Algorithm 4): https://arxiv.org/abs/1805.10941
+ * The 64-bit version of the nearly divisionless algorithm requires 128-bit
+ * integer arithmetic. Instead of trying to deal with cross-platform support for
+ * `__int128`, fallback to using the Java algorithm for 64-bit instead. */
+static uint64_t
+_mongoc_rand_java64 (uint64_t s, uint64_t (*rand64) (void))
+{
+   const uint64_t limit = UINT64_MAX; /* 2^L */
+   uint64_t x, r;
+
+   x = rand64 ();
+   r = x % s;
+
+   while ((x - r) > (limit - s)) {
+      x = rand64 ();
+      r = x % s;
+   }
+
+   return r;
+}
+
+#if defined(MONGOC_ENABLE_CRYPTO)
+/* Attempt to use `_mongoc_rand_bytes` if available. */
```

Comment:
What is the motivation for preferring `_mongoc_rand_bytes` instead of `rand`?

One concern I have is that is mixes intention. Callers of `_mongoc_rand_uint32_t`, `_mongoc_rand_uint64_t`, and `_mongoc_rand_size_t` are not guaranteed a random number derived from a secure random source depending on the platform. 

Secure random generation may also be significantly slower on some platforms.